src/playlist-maker.js

```diff
@@ -428,26 +453,43 @@ export default function factory(player, initialList, initialIndex = 0) {
   /**
    * Shuffle the contents of the list randomly.
    *
-   * @see {@link https://github.com/lodash/lodash/blob/40e096b6d5291a025e365a0f4c010d9a0efb9a69/shuffle.js}
+   * @see   {@link https://github.com/lodash/lodash/blob/40e096b6d5291a025e365a0f4c010d9a0efb9a69/shuffle.js}
    * @fires playlistsorted
+   * @todo  Make the `rest` option default to `true` in v5.0.0.
+   * @param {Object} [options]
+   *        An object containing shuffle options.
+   *
+   * @param {boolean} [options.rest = false]
+   *        By default, the entire playlist is randomized. However, this may
+   *        not be desirable in all cases, such as when a user is already
+   *        watching a video.
+   *
+   *        When `true` is passed for this option, it will only shuffle
+   *        playlist items after the current item. For example, when on the
+   *        first item, will shuffle the second item and beyond.
    */
-  playlist.shuffle = () => {
-    let index = -1;
-    const length = list.length;
+  playlist.shuffle = ({rest} = {}) => {
+    let index = 0;
+    let arr = list;
+
+    // When options.rest is true, start randomization at the item after the
+    // current item.
+    if (rest) {
+      index = playlist.currentIndex_ + 1;
+      arr = list.slice(index);
+    }
 
-    // Bail if the array is empty.
-    if (!length) {
+    // Bail if the array is empty or too short to shuffle.
+    if (arr.length <= 1) {
       return;
     }
 
-    const lastIndex = length - 1;
-
-    while (++index < length) {
-      const rand = index + Math.floor(Math.random() * (lastIndex - index + 1));
-      const value = list[rand];
+    randomize(arr);
 
-      list[rand] = list[index];
-      list[index] = value;
+    // When options.rest is true, splice the randomized sub-array back into
+    // the original array.
+    if (rest) {
+      list.splice(...[index, arr.length].concat(arr));
```

Comment:
this is probably incredibly slow, though, hopefully, not that big of an issue for most usecases

Reply:
Possibly, but likely only if someone is calling `shuffle` often. We can optimize if it becomes an issue.